middle_end/flambda2/to_cmm/to_cmm_expr.ml

```diff
@@ -460,123 +368,158 @@ and let_cont_inline env res k h body =
   (* Translate each continuation handler *)
-  let map, res =
+  let conts_to_handlers, res =
     Continuation.Map.fold
-      (fun k h (map, res) ->
-        let vars, _arity, handler, res = continuation_handler env res h in
-        Continuation.Map.add k (vars, handler) map, res)
-      map
+      (fun k handler (conts_to_handlers, res) ->
+        let vars, _arity, handler, res = continuation_handler env res handler in
+        Continuation.Map.add k (vars, handler) conts_to_handlers, res)
+      conts_to_handlers
       (Continuation.Map.empty, res)
   in
   let dbg = Debuginfo.none (* CR mshinwell: fix debuginfo *) in
-  (* Setup the cmm handlers for the static catch *)
+  (* Setup the Cmm handlers for the Ccatch *)
   let handlers =
     Continuation.Map.fold
-      (fun k (vars, handle) acc ->
+      (fun k (vars, handler) acc ->
         let id = Env.get_cmm_continuation env k in
-        C.handler ~dbg id vars handle :: acc)
-      map []
+        C.handler ~dbg id vars handler :: acc)
+      conts_to_handlers []
   in
   let body, res = expr env res body in
   wrap (C.create_ccatch ~rec_flag:true ~body ~handlers), res
 
-and continuation_handler env res h =
-  Continuation_handler.pattern_match' h
+and continuation_handler env res handler =
+  Continuation_handler.pattern_match' handler
     ~f:(fun params ~num_normal_occurrences_of_params:_ ~handler ->
       let arity = Bound_parameters.arity params in
       let env, vars = C.bound_parameters env params in
-      let e, res = expr env res handler in
-      vars, arity, e, res)
+      let expr, res = expr env res handler in
+      vars, arity, expr, res)
+
+and apply_expr env res apply =
+  let call, env, effs = translate_apply env apply in
+  (* With respect to flushing the environment we have three cases:
 
-(* Function calls: besides the function calls, there are a few things to do:
+     1. The call never returns or jumps to another function
 
-   - setup the mutable variables for the exn cont extra args if needed
+     2. The call jumps to somewhere else in the current function, but there is
+     more than one incoming control flow edge to that point
 
-   - translate the call continuation (either through a jump, or inlining). *)
-and apply_expr env res e =
-  let call, env, effs = apply_call env e in
-  let k_exn = Apply.exn_continuation e in
-  let dbg = Apply.dbg e in
-  let call, env = wrap_call_exn ~dbg env e call k_exn in
-  match Apply.continuation e with
+     3. The call jumps to somewhere else in the current function and is the only
+     thing that jumps to that point.
+
+     In cases 1 and 2 we can't affect the code generation at the jump target
+     (either because it wasn't generated by us, or is in another function) so
+     the environment must be flushed prior to the call. In case 3 we can avoid
+     flushing the environment due to the linearity of the control flow and the
+     fact that we are able to control the code generation at the jump target. *)
```

Comment:
I don't understand this comment (maybe some examples for each behaviour would help understand), and I don't see what points it tries to make and where in the body  of `apply_expr` it changes things. For isntance, point 3 seems to refer to situations where we inline the return continuation of the call, but if so, it'd be best to explicit state that (and not leave it to reader to guess).

Overall, I feel the comment is slightly off in the way it present things. The reason why some cases flush and other don't mainly depend on one thing, which is whether we directly return some cmm code, or we recursively call a translation function:
- if we directly return cmm code, then we have to flush, because the env is effectively dropped, and thus without flushing, any binding in it would be lost (leading to variables used but not defined)
- in the case where we recurse, we pass the env to the `expr` translation function, which we trust to flush the env at the end.